torch/distributed/_fsdp/fully_sharded_data_parallel.py

```diff
@@ -0,0 +1,803 @@
+    to make sure APIs are called in the correct state.
+    ..note::
+        BACKWARD_PRE and BACKWARD_POST states are used to ensure we
+        receives backward hooks in the correct order. It is used to catch
+        unexpected order of hooks being called (likely due to our
+        hook registration logic or autograd engine logic changes).
+    """
+
+    IDLE = auto()
+    FORWARD = auto()
+    BACKWARD_PRE = auto()
+    BACKWARD_POST = auto()
+
+
+class FullyShardedDataParallel(nn.Module):
+    """
+    A wrapper for sharding Module parameters across data parallel workers. This
+    is inspired by `Xu et al.`_ as well as the ZeRO Stage 3 from DeepSpeed_.
+    FullyShardedDataParallel is commonly shorten to FSDP.
+    .. _`Xu et al.`: https://arxiv.org/abs/2004.13336
+    .. _DeepSpeed: https://www.deepspeed.ai/
+    Example::
+        import torch
+        from torch.distributed._fsdp import FullyShardedDataParallel as FSDP
+        torch.cuda.set_device(device_id)
+        sharded_module = FSDP(my_module)
+        optim = torch.optim.Adam(sharded_module.parameters(), lr=0.0001)
+        x = sharded_module(x, y=3, z=torch.Tensor([1]))
+        loss = x.sum()
+        loss.backward()
+        optim.step()
+    .. warning::
+        The optimizer must be initialized *after* the module has been wrapped,
+        since FSDP will shard parameters in-place and this will break any
+        previously initialized optimizers.
+    Args:
+        module (nn.Module):
+            module to be wrapped with FSDP.
+        process_group (Optional):
+            process group for sharding
+    """
+
+    def __init__(
+        self,
+        module: nn.Module,
+        process_group: Optional[ProcessGroup] = None,
+    ):
+        super().__init__()
+        self.module = module
+        self.process_group = process_group or _get_default_group()
+        self.rank = self.process_group.rank()
+        self.world_size = self.process_group.size()
+        # device for computation, if module is on GPU, use module.device;
+        # if module is on CPU, use current device;
+        self.compute_device = _get_default_cuda_device(module)
+        self.compute_dtype = next(module.parameters()).dtype
+
+        # Free full params and keep shard only after forward
+        self.reshard_after_forward = True
+
+        # setting two factors to avoid underflow and overflow
+        self.gradient_predivide_factor: float = self._get_gradient_predivide_factor(self.world_size)
+        self.gradient_postdivide_factor: float = self.world_size / self.gradient_predivide_factor
+
+        self.numel_padded_per_param: List[int] = []
+
+        # Only handle params which are not already sharded. This enables
+        # sharding individual layers of a Module, with an outer wrapper to
+        # shard any leftover parameters.
+        params = []
+        for param_name, param in module.named_parameters():
+            if not hasattr(param, "_is_sharded"):
```

Comment:
What could be setting such a flag in core?

Reply:
it is only set in FSDP after it is sharded, let me take @mrshenli's suggestion to make it like "param.fsdp_attr._is_sharded"

Reply:
So this case would only happen if you wrap an FSDP model into another FSDP model? Is that allowed?